node/src/components/linear_chain.rs

```diff
@@ -286,60 +335,135 @@ where
                     public_key,
                     ..
                 } = *fs;
-                // TODO: Also verify that the public key belongs to a bonded validator!
                 if let Err(err) = fs.verify() {
                     warn!(%block_hash, %public_key, %err, "received invalid finality signature");
                     return Effects::new();
                 }
-                debug!(%block_hash, %public_key, "received new finality signature");
-                let sigs = self
-                    .pending_finality_signatures
-                    .entry(public_key)
-                    .or_default();
-                // Limit the memory we use for storing unknown signatures from each validator.
-                if sigs.len() >= MAX_PENDING_FINALITY_SIGNATURES_PER_VALIDATOR {
-                    warn!(
-                        %block_hash, %public_key,
-                        "received too many finality signatures for unknown blocks"
-                    );
-                    return Effects::new();
-                }
-                // Add the pending signature and request the block from storage.
-                if sigs.insert(block_hash, *fs).is_some() {
-                    return Effects::new(); // Signature was already known.
-                }
                 effect_builder
                     .get_block_from_storage(block_hash)
                     .event(move |maybe_block| {
                         let maybe_box_block = maybe_block.map(Box::new);
-                        Event::GetBlockForFinalitySignaturesResult(block_hash, maybe_box_block)
+                        Event::GetBlockForFinalitySignaturesResult(fs, maybe_box_block)
                     })
             }
-            Event::GetBlockForFinalitySignaturesResult(block_hash, None) => {
-                let signers = self
-                    .pending_finality_signatures
-                    .values()
-                    .filter_map(|sigs| sigs.get(&block_hash))
-                    .map(|fs| format!("{}", fs.public_key))
-                    .collect_vec();
-                if !signers.is_empty() {
-                    // We have signatures for an unknown block. Print log messages.
-                    warn!(
-                       %block_hash, signers = %signers.join(", "),
-                       "received signatures for a block that was not found in storage"
+            Event::GetBlockForFinalitySignaturesResult(fs, maybe_block) => {
+                if let Some(block) = &maybe_block {
+                    assert_eq!(
```

Comment:
We should also check whether `fs.era_id() == block.header().era_id()`.

Reply:
In this case it wouldn't be an `assert`ion though but an invalid signature,right?

Reply:
Yes, you're right.

Reply:
…and we need to have the same check in the other case, where we receive the block later than the signature. (Or does that end up in the same place in the code?)

Reply:
No, it does not end up in the same place in the code. Good point.

Reply:
Done in [`3bffb83` (#683)](https://github.com/CasperLabs/casper-node/pull/683/commits/3bffb831458819c2ed21e895cd3fcf945c06f1a4)